handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java

```diff
@@ -0,0 +1,746 @@
+import io.netty.util.ReferenceCounted;
+import io.netty.util.ResourceLeak;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakDetectorFactory;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.StringUtil;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import org.apache.tomcat.jni.CertificateVerifier;
+import org.apache.tomcat.jni.Pool;
+import org.apache.tomcat.jni.SSL;
+import org.apache.tomcat.jni.SSLContext;
+
+import java.security.AccessController;
+import java.security.PrivateKey;
+import java.security.PrivilegedAction;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.CertificateRevokedException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509ExtendedKeyManager;
+import javax.net.ssl.X509ExtendedTrustManager;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * An implementation of {@link SslContext} which works with libraries that support the
+ * <a href="https://www.openssl.org/">OpenSsl</a> C library API.
+ * <p>Instances of this class must be {@link #release() released} or else native memory will leak!
+ */
+public abstract class ReferenceCountedOpenSslContext extends SslContext implements ReferenceCounted {
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(ReferenceCountedOpenSslContext.class);
+    /**
+     * To make it easier for users to replace JDK implemention with OpenSsl version we also use
+     * {@code jdk.tls.rejectClientInitiatedRenegotiation} to allow disabling client initiated renegotiation.
+     * Java8+ uses this system property as well.
+     * <p>
+     * See also <a href="http://blog.ivanristic.com/2014/03/ssl-tls-improvements-in-java-8.html">
+     * Significant SSL/TLS improvements in Java 8</a>
+     */
+    private static final boolean JDK_REJECT_CLIENT_INITIATED_RENEGOTIATION =
+            SystemPropertyUtil.getBoolean("jdk.tls.rejectClientInitiatedRenegotiation", false);
+    private static final List<String> DEFAULT_CIPHERS;
+    private static final Integer DH_KEY_LENGTH;
+    private static final ResourceLeakDetector<ReferenceCountedOpenSslContext> leakDetector =
+            ResourceLeakDetectorFactory.instance().newResourceLeakDetector(ReferenceCountedOpenSslContext.class);
+
+    // TODO: Maybe make configurable ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)